src/dwarf/debug_info.rs

```diff
@@ -221,336 +230,140 @@ pub enum AttrValue<'a> {
 ///
 /// Return AttrValue and the number of bytes it takes.
 fn extract_attr_value(
-    data: &[u8],
+    mut data: &[u8],
     form: u8,
     dwarf_sz: usize,
     addr_sz: usize,
-) -> Option<(AttrValue, usize)> {
+) -> Option<(AttrValue<'_>, usize)> {
     match form {
         DW_FORM_addr => {
             if addr_sz == 0x4 {
-                if 4 <= data.len() {
-                    Some((AttrValue::Unsigned(decode_uword(data) as u64), 4))
-                } else {
-                    None
-                }
-            } else if 8 <= data.len() {
-                Some((AttrValue::Unsigned(decode_udword(data)), 8))
+                Some((AttrValue::Unsigned(data.read_u32()? as u64), 4))
             } else {
-                None
+                Some((AttrValue::Unsigned(data.read_u64()?), 8))
             }
         }
         DW_FORM_block2 => {
-            if 2 <= data.len() {
-                let bytes = decode_uhalf(data);
-                let fullsize = bytes as usize + 2;
-                if fullsize <= data.len() {
-                    Some((AttrValue::Bytes(&data[2..fullsize]), fullsize))
-                } else {
-                    None
-                }
-            } else {
-                None
-            }
+            let bytes = data.read_u16()?;
+            let fullsize = bytes as usize + 2;
+            Some((AttrValue::Bytes(data.read_slice(bytes.into())?), fullsize))
         }
         DW_FORM_block4 => {
-            if 2 <= data.len() {
-                let bytes = decode_uword(data);
-                let fullsize = bytes as usize + 2;
-                if fullsize <= data.len() {
-                    Some((AttrValue::Bytes(&data[2..fullsize]), fullsize))
-                } else {
-                    None
-                }
-            } else {
-                None
-            }
-        }
-        DW_FORM_data2 => {
-            if 2 <= data.len() {
-                let v = decode_uhalf(data);
-                Some((AttrValue::Unsigned(v as u64), 2))
-            } else {
-                None
-            }
-        }
-        DW_FORM_data4 => {
-            if 4 <= data.len() {
-                let v = decode_uword(data);
-                Some((AttrValue::Unsigned(v as u64), 4))
-            } else {
-                None
-            }
-        }
-        DW_FORM_data8 => {
-            if 8 <= data.len() {
-                let v = decode_udword(data);
-                Some((AttrValue::Unsigned(v), 8))
-            } else {
-                None
-            }
+            let bytes = data.read_u32()?;
+            let fullsize = bytes as usize + 2;
+            Some((AttrValue::Bytes(data.read_slice(bytes as usize)?), fullsize))
```

Comment:
above you used `bytes.into()` (which gave me a pause for a bit :) ), should we use one consistent form? BTW, which one is more "idiomatic": explicit `as` conversion, or `.into()`?

Reply:
I believe we should use `.into()` where possible. The reason is that `.into()` is guaranteed to be a non-fallible conversion by design (and so that is conveyed to the reader). `as` can truncate and generally introduce bugs in the logic because of that, so generally warrants some more thinking.

Unfortunately, `u32` is not guaranteed to be representable by `usize` (similar to the relationship between `uint32_t` and `size_t` in C) and so there is no guaranteed infallible conversion (but there is for `u16`). We can use `try_into().unwrap()` (or bubble up the error, of course). Given the existing `as` cast in that very context and unlikeliness of `u32` not actually fitting into `usize` on any common system (I believe), I didn't do that. Not a strong reason by any means and happy to change it.

Reply:
Makes sense, thanks for explaining! try_into().unwrap() seems to be an overkill, let's keep `as usize` as is